### PR TITLE
95 notebooks

### PR DESCRIPTION
**Pull request by anschaible**
This branch is based on the final-load-gas-particles. Once the final-load-gas-particles is on the main, this branch should alos be merged to the main.

This branch contains a cleaned version of the notebooks with many comments to explain the notebooks. Only missing thing is the config for the ssp_templates_fsps.ipynb. I want use some of these notebooks for the documentaion.

### CODE REVIEW

File: doc/conf.py [hunk 1]
```diff
@@ -7,9 +7,9 @@
 
 # -- Project information -----------------------------------------------------
 
-project = 'rubix'
-copyright = '2024, Ufuk Çakır'
-author = 'Ufuk Çakır'
+project = "rubix"
+copyright = "2024, Ufuk Çakır"
+author = "Ufuk Çakır"
```
Comment (TobiBu) [inline doc/conf.py:12]: By now: can we add more authors here @ufuk-cakir?

**Review by TobiBu — APPROVED** [review]

File: doc/conf.py (hunk 1 above)
Comment (TobiBu) [reply doc/conf.py:12]: and if you have time to approve this PR, we can merge :-)

**Comment by anschaible** [comment]
@MaHaWo currently all tests are failing, we assume it has something to do with the new jax version 0.3.6, because GitHub uses 0.3.6 and it breakes and locally I use 0.3.5 and everything works. @MaHaWo could you have a look into it?

File: rubix/core/data.py [hunk 2]
```diff
@@ -10,6 +12,118 @@
...
+    def tree_flatten(self):
+        children = (self.redshift, self.center, self.halfmassrad_stars)
+        aux_data = {}
+        return children, aux_data
+
+    @classmethod
+    def tree_unflatten(cls, aux_data, children):
+        return cls(*children)
+
+@partial(jax.tree_util.register_pytree_node_class)
```
Comment (ufuk-cakir) [inline rubix/core/data.py:35]: we should have a Look if we want to use [equinox module](https://docs.kidger.site/equinox/api/module/module/) or [chex dataclass](https://chex.readthedocs.io/en/latest/api.html#chex.dataclass) and what the differences are

Comment (ufuk-cakir) [reply rubix/core/data.py:35]: I think this automatically impelements the class as a pytree

File: rubix/core/noise.py [hunk 3]
```diff
@@ -29,12 +29,11 @@ def get_apply_noise(config: dict):
 
     logger = get_logger()
 
-    def apply_noise(inputs: dict[str, jax.Array]) -> dict[str, jax.Array]:
+    def apply_noise(rubixdata: object) -> object:
```
Comment (ufuk-cakir) [inline rubix/core/noise.py:32]: I think I wrote that somewhere else, but instead of `object` we could create a type for `RubixData` or import the class here and explicitly add that to the typing argument

File: rubix/core/rotation.py [hunk 4]
```diff
@@ -43,29 +43,57 @@ def get_galaxy_rotation(config: dict):
...
-            alpha=alpha,
-            beta=beta,
-            gamma=gamma,
-        )
-
-        # Update the inputs
-        inputs["coords"] = coords
-        inputs["velocities"] = velocities
-        return inputs
+        if "stars" in config["data"]["args"]["particle_type"]:
```
Comment (ufuk-cakir) [inline rubix/core/rotation.py:49]: is this old code?

I feel like we had a imprvoved version on this where we refactored this into different funciont, since this is just copy past of almost the same code for both stars and gas.

we should refactor this

File: rubix/galaxy/input_handler/illustris.py [hunk 5]
```diff
@@ -54,9 +54,75 @@ def get_units(self):
...
+                raise NotImplementedError(
+                    f"Unsupported particle type found in the file: {field}"
+                )
+            else:
+                raise ValueError(f"Unexpected fields found in the file: {extra_fields}")
+
+        self._logger.debug(f"Matching fields: {matching_fields}")
+
+    """
+    def _check_particle_data(self, particle_data, units):
```
Comment (ufuk-cakir) [inline rubix/galaxy/input_handler/illustris.py:83]: why is this commented out?

**Review by ufuk-cakir — CHANGES_REQUESTED** [review]

File: rubix/core/data.py (hunk 2 above)
Comment (anschaible) [reply rubix/core/data.py:35]: @ufuk-cakir the same discussion is ongoing on the branch final load gas particles

Comment (anschaible) [reply rubix/core/data.py:35]: equinox has a problem, if some values in the beginning are None and then set to a jax.array

Comment (anschaible) [reply rubix/core/data.py:35]: We can change the initialisation of the class. @ufuk-cakir could you please provide a suggestion how to change the code?

File: rubix/core/noise.py (hunk 3 above)
Comment (anschaible) [reply rubix/core/noise.py:32]: Yes, this is now implemented on the documentation-sphinx branch

File: rubix/core/rotation.py (hunk 4 above)
Comment (anschaible) [reply rubix/core/rotation.py:49]: Yes, this is old code, the most up to date version is currently on the branch final load gas particles

File: rubix/galaxy/input_handler/illustris.py (hunk 5 above)
Comment (anschaible) [reply rubix/galaxy/input_handler/illustris.py:83]: Not sure, why it was commented out, now I included the part again

File: rubix/core/data.py (hunk 2 above)
Comment (ufuk-cakir) [reply rubix/core/data.py:58]: what about instead of dataclass we use python native `NamedTuple` ? It looks like there is not really any functionality defined for the dataclass, but rather used to define which variables the data structure holds. As far as I know, JAX works with NamedTuples, and gradient computation should also work.

Comment (anschaible) [reply rubix/core/data.py:58]: I think the notebook branch is the wrong branch to do large changes on the dataclass of the rubixdata. If you want to change it, please open a new issue and branch, otherwise it gets a too large mess on the individual branches and it is difficult to understand the changes

Comment (anschaible) [reply rubix/core/data.py:58]: If we have the spectra or datacube as field in the rubixdata, this is when initializing the dataclass not peresent at that point, because it is calculated later in the pipeline

**Review by TobiBu — APPROVED** [review]

File: rubix/core/data.py (hunk 2 above)
Comment (TobiBu) [reply rubix/core/data.py:58]: > what about instead of dataclass we use python native `NamedTuple` ? It looks like there is not really any functionality defined for the dataclass, but rather used to define which variables the data structure holds. As far as I know, JAX works with NamedTuples, and gradient computation should also work.

also moved this to [issue](https://github.com/ufuk-cakir/rubix/issues/119) so we have a reference.
Can we merge, so we can progress with all other PRs?

**Review by TobiBu — APPROVED** [review]

File: rubix/core/data.py [hunk 6]
```diff
@@ -37,6 +37,29 @@
 #        return self.attributes
 
 
+# class Particles:
```
Comment (ufuk-cakir) [inline rubix/core/data.py:40]: why is this commented out?
